src/inmanta/data/__init__.py

```diff
@@ -1703,6 +1703,80 @@ async def purge_logs(cls):
             value = cls._get_value(keep_logs_until)
             await cls._execute_query(query, value)
 
+    @classmethod
+    async def query_resource_actions(
+        cls,
+        environment: uuid.UUID,
+        resource_type: Optional[str] = None,
+        agent: Optional[str] = None,
+        attribute: Optional[str] = None,
+        attribute_value: Optional[str] = None,
+        log_severity: Optional[str] = None,
+        limit: int = 0,
+        first_timestamp: Optional[datetime.datetime] = None,
+        last_timestamp: Optional[datetime.datetime] = None,
```

Comment:
Doing pagination based on the timestamp of a ResourceAction is not going to work, because multiple ResourceActions may have the same timestamp. Imagine R1 and R2 are ResourceActions with the same timestamp and R1 appears at the end of page1. In that case R2 will not be displayed when page2 is requested from the server.

Reply:
That's true. Do you have any suggestions? 

My first ideas (using both first and last as inclusive limits for the range; using the `version` values instead of the timestamps; adding the `action_id` to the ordering; specifying the only the start and end without `limit` - or interpreting the `limit` as an interval) either don't solve the problem or lead to different issues.

Reply:
The best solution I see is adding the `action_id` column to the ordering. Unfortunately, this will make the API quite ugly, but I don't see a better solution.